File: packages/go/qorechain/query/jsonrpc.go
package query

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"sync"
)

// JSONRPCError is returned when a JSON-RPC response carries an error member.
type JSONRPCError struct {
	Code    int             `json:"code"`
	Message string          `json:"message"`
	Data    json.RawMessage `json:"data,omitempty"`
}

func (e *JSONRPCError) Error() string {
	return e.Message
}

type jsonrpcRequest struct {
	JSONRPC string `json:"jsonrpc"`
	ID      int    `json:"id"`
	Method  string `json:"method"`
	Params  []any  `json:"params"`
}

type jsonrpcResponse struct {
	Result json.RawMessage `json:"result"`
	Error  *JSONRPCError   `json:"error"`
}

// JSONRPCClient is a minimal JSON-RPC 2.0 client over HTTP POST. Request ids
// auto-increment per client.
type JSONRPCClient struct {
	url    string
	http   *http.Client
	mu     sync.Mutex
	nextID int
}

// NewJSONRPCClient creates a JSON-RPC client targeting the given URL. If
// httpClient is nil, http.DefaultClient is used.
func NewJSONRPCClient(rpcURL string, httpClient *http.Client) *JSONRPCClient {
	if httpClient == nil {
		httpClient = http.DefaultClient
	}
	return &JSONRPCClient{url: rpcURL, http: httpClient, nextID: 1}
}

// Call invokes a JSON-RPC method and returns its result. It returns a
// *JSONRPCError when the response contains an error member, or an *HTTPError on
// a non-2xx transport response.
func (c *JSONRPCClient) Call(method string, params []any) (json.RawMessage, error) {
	c.mu.Lock()
	id := c.nextID
	c.nextID++
	c.mu.Unlock()

	if params == nil {
		params = []any{}
	}
	reqBody, err := json.Marshal(jsonrpcRequest{JSONRPC: "2.0", ID: id, Method: method, Params: params})
	if err != nil {
		return nil, err
	}
	req, err := http.NewRequest(http.MethodPost, c.url, bytes.NewReader(reqBody))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", defaultUserAgent)

	resp, err := c.http.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return nil, &HTTPError{Status: resp.StatusCode, URL: c.url, Body: string(body)}
	}
	var parsed jsonrpcResponse
	if err := json.Unmarshal(body, &parsed); err != nil {
		return nil, fmt.Errorf("invalid JSON-RPC response: %w", err)
	}
	if parsed.Error != nil {
		return nil, parsed.Error
	}
	return parsed.Result, nil
}

File: packages/py/src/qorechain/jsonrpc.py
"""Generic JSON-RPC 2.0 clients (sync + async).

Manage request ids (auto-incrementing per client), serialize the standard
``{"jsonrpc", "id", "method", "params"}`` envelope, and map JSON-RPC error
responses to a typed :class:`JsonRpcError`. These are the transport base for the
QoreChain ``qor_*`` namespace and any EVM ``eth_*`` methods.
"""

from __future__ import annotations

from typing import Any

import httpx

from .rest import QoreHttpError

_DEFAULT_TIMEOUT = 30.0


class JsonRpcError(Exception):
    """Raised when a JSON-RPC response carries an ``error`` member."""

    def __init__(self, code: int, message: str, data: Any = None) -> None:
        super().__init__(message)
        self.code = code
        self.data = data


def _build_body(next_id: int, method: str, params: list[Any] | None) -> dict[str, Any]:
    return {"jsonrpc": "2.0", "id": next_id, "method": method, "params": params or []}


def _parse_response(resp: httpx.Response, url: str) -> Any:
    if resp.status_code < 200 or resp.status_code >= 300:
        raise QoreHttpError(resp.status_code, url, resp.text)
    payload = resp.json()
    err = payload.get("error")
    if err:
        raise JsonRpcError(err.get("code", 0), err.get("message", ""), err.get("data"))
    return payload.get("result")


class JsonRpcClient:
    """A minimal synchronous JSON-RPC 2.0 client over HTTP POST."""

    def __init__(
        self,
        url: str,
        *,
        timeout: float = _DEFAULT_TIMEOUT,
        client: httpx.Client | None = None,
        headers: dict[str, str] | None = None,
    ) -> None:
        self.url = url
        self._owns_client = client is None
        self._client = client or httpx.Client(timeout=timeout, headers=headers)
        self._next_id = 1

    def close(self) -> None:
        if self._owns_client:
            self._client.close()

    def __enter__(self) -> JsonRpcClient:
        return self

    def __exit__(self, *exc: object) -> None:
        self.close()

    def call(self, method: str, params: list[Any] | None = None) -> Any:
        """Invoke a JSON-RPC method and return its ``result``.

        :raises JsonRpcError: When the response contains an ``error``.
        :raises QoreHttpError: On a non-2xx transport response.
        """
        body = _build_body(self._next_id, method, params)
        self._next_id += 1
        resp = self._client.post(self.url, json=body)
        return _parse_response(resp, self.url)


class AsyncJsonRpcClient:
    """Asynchronous mirror of :class:`JsonRpcClient`."""

    def __init__(
        self,
        url: str,
        *,
        timeout: float = _DEFAULT_TIMEOUT,
        client: httpx.AsyncClient | None = None,
        headers: dict[str, str] | None = None,
    ) -> None:
        self.url = url
        self._owns_client = client is None
        self._client = client or httpx.AsyncClient(timeout=timeout, headers=headers)
        self._next_id = 1

    async def aclose(self) -> None:
        if self._owns_client:
            await self._client.aclose()

    async def __aenter__(self) -> AsyncJsonRpcClient:
        return self

    async def __aexit__(self, *exc: object) -> None:
        await self.aclose()

    async def call(self, method: str, params: list[Any] | None = None) -> Any:
        """Invoke a JSON-RPC method and return its ``result``."""
        body = _build_body(self._next_id, method, params)
        self._next_id += 1
        resp = await self._client.post(self.url, json=body)
        return _parse_response(resp, self.url)
